wgengine/magicsock/magicsock.go

```diff
@@ -5105,6 +5125,88 @@ func (s derpAddrFamSelector) PreferIPv6() bool {
 	return false
 }
 
+type endpointTrackerEntry struct {
+	endpoint tailcfg.Endpoint
+	until    time.Time
+}
+
+type endpointTracker struct {
+	mu    sync.Mutex
+	cache map[netip.AddrPort]endpointTrackerEntry
+}
+
+const endpointTrackerLifetime = 5*time.Minute + 10*time.Second // slightly longer than a full netcheck period
+
+func (et *endpointTracker) update(now time.Time, eps []tailcfg.Endpoint) (ret []tailcfg.Endpoint) {
+	ret = eps
+
+	// We can have a very large set of endpoints for machines with many
+	// interfaces; rather than a O(n²)-ish loop over every entry in eps for
+	// each entry in the cache, create a map so that lookups are cheap.
+	//
+	// TODO(andrew): we could probably have a fast path for when we have a
+	// small number of endpoints.
+	inputEps := make(map[netip.AddrPort]bool, len(eps))
+	for _, ep := range eps {
+		inputEps[ep.Addr] = true
+	}
```

Comment:
yeah, I guess the only operation you care about is "is it in the set?" Our corp `misc/set.Slice` basically does this with its:

```go
// Slice is a set of elements tracked in a slice of unique elements.
type Slice[T comparable] struct {
	slice []T
	set   map[T]bool // nil until/unless slice is large enough
}

// Contains reports whether v is in the set.
// The amortized cost is O(1).
func (ss *Slice[T]) Contains(v T) bool {
	if ss.set != nil {
		return ss.set[v]
	}
	return slices.Index(ss.slice, v) != -1
}
```

Reply:
Moved that type here to oss and pushed an update to use it 👍